src/theory/quantifiers/quantifiers_rewriter.cpp

```diff
@@ -196,60 +196,75 @@ void QuantifiersRewriter::computeArgVec2(const std::vector<Node>& args,
-  if( in.getKind()==kind::EXISTS || in.getKind()==kind::FORALL ){
-    Trace("quantifiers-rewrite-debug") << "pre-rewriting " << in << std::endl;
-    std::vector< Node > args;
-    Node body = in;
-    bool doRewrite = false;
-    while( body.getNumChildren()==2 && body.getKind()==body[1].getKind() ){
-      for( unsigned i=0; i<body[0].getNumChildren(); i++ ){
-        args.push_back( body[0][i] );
-      }
-      body = body[1];
-      doRewrite = true;
-    }
-    if( doRewrite ){
-      std::vector< Node > children;
-      for( unsigned i=0; i<body[0].getNumChildren(); i++ ){
-        args.push_back( body[0][i] );
-      }      
-      children.push_back( NodeManager::currentNM()->mkNode(kind::BOUND_VAR_LIST,args) );
-      children.push_back( body[1] );
-      if( body.getNumChildren()==3 ){
-        children.push_back( body[2] );
-      }
-      Node n = NodeManager::currentNM()->mkNode( in.getKind(), children );
-      if( in!=n ){
-        Trace("quantifiers-pre-rewrite") << "*** pre-rewrite " << in << std::endl;
-        Trace("quantifiers-pre-rewrite") << " to " << std::endl;
-        Trace("quantifiers-pre-rewrite") << n << std::endl;
-      }
-      return RewriteResponse(REWRITE_DONE, n);
-    }
-  }
   return RewriteResponse(REWRITE_DONE, in);
 }
 
-RewriteResponse QuantifiersRewriter::postRewrite(TNode in) {
+RewriteResponse QuantifiersRewriter::postRewrite(TNode in)
+{
   Trace("quantifiers-rewrite-debug") << "post-rewriting " << in << std::endl;
   RewriteStatus status = REWRITE_DONE;
   Node ret = in;
   RewriteStep rew_op = COMPUTE_LAST;
-  //get the body
-  if( in.getKind()==EXISTS ){
-    std::vector< Node > children;
-    children.push_back( in[0] );
-    children.push_back( in[1].negate() );
-    if( in.getNumChildren()==3 ){
-      children.push_back( in[2] );
+  // get the body
+  if (in.getKind() == EXISTS)
+  {
+    std::vector<Node> children;
+    children.push_back(in[0]);
+    children.push_back(in[1].negate());
+    if (in.getNumChildren() == 3)
+    {
+      children.push_back(in[2]);
     }
-    ret = NodeManager::currentNM()->mkNode( FORALL, children );
+    ret = NodeManager::currentNM()->mkNode(FORALL, children);
     ret = ret.negate();
     status = REWRITE_AGAIN_FULL;
-  }else if( in.getKind()==FORALL ){
-    if( in[1].isConst() && in.getNumChildren()==2 ){
+  }
+  else if (in.getKind() == FORALL)
+  {
+    std::vector<Node> args;
+    Node body = in;
+    bool combineQuantifiers = false;
+    bool continueCombine = false;
+    do
+    {
+      for (const Node& v : body[0])
+      {
+        if (std::find(args.begin(), args.end(), v)==args.end())
+        {
+          args.push_back(v);
+        }
+      }
+      if (body.getNumChildren() == 2 && body.getKind() == body[1].getKind())
+      {
+        body = body[1];
+        continueCombine = true;
+        combineQuantifiers = true;
+      }
+      else
+      {
+        continueCombine = false;
+      }
+    }
+    while (continueCombine);
```

Comment:
the condition could be `while (true)` if rather than the else above you just had a `break` (and a `continue` when you are continuing to combine). Then you could also remove this `continueCombine` variable.